src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts

```diff
@@ -0,0 +1,742 @@
+    // texture with origin (_, 1, 1) (in texel blocks). Similarly for other coordinates.
+    t.testRun({
+      textureDataLayout: { offset: 0, bytesPerRow, rowsPerImage },
+      copySize,
+      dataSize,
+      origin,
+      textureSize,
+      format,
+      initMethod,
+      checkMethod,
+    });
+  });
+
+// Generates textureSizes which correspond to the same physicalSizeAtMipLevel
+// including virtual sizes at mip level different from the physical ones.
+function* textureSizeExpander({
+  format,
+  mipLevel,
+  texturePhysicalSizeAtMipLevelInBlocks,
+}: {
+  format: SizedTextureFormat;
+  mipLevel: number;
+  texturePhysicalSizeAtMipLevelInBlocks: GPUExtent3DDict;
+}): Generator<{ textureSize: [number, number, number] }> {
+  const info = kSizedTextureFormatInfo[format];
+  const textureSize = [
+    (texturePhysicalSizeAtMipLevelInBlocks.width * info.blockWidth!) << mipLevel,
+    (texturePhysicalSizeAtMipLevelInBlocks.height * info.blockHeight!) << mipLevel,
+    texturePhysicalSizeAtMipLevelInBlocks.depth,
+  ];
+  yield {
+    textureSize: textureSize as [number, number, number],
+  };
+
+  // We choose width and height of the texture so that the values are divisible by blockWidth and
+  // blockHeight respectively and so that the virtual size at mip level corresponds to the same
+  // physical size.
+  // Virtual size at mip level with modified width has width = (physical size width) - (blockWidth / 2).
+  // Virtual size at mip level with modified height has height = (physical size height) - (blockHeight / 2).
+  assert(mipLevel > 0);
+  assert(textureSize[0] > 0 && textureSize[1] > 0);
+  const modifiedWidth = textureSize[0] - (info.blockWidth! << (mipLevel - 1));
+  const modifiedHeight = textureSize[1] - (info.blockHeight! << (mipLevel - 1));
+
+  if (info.blockWidth! > 1 && modifiedWidth !== textureSize[0]) {
+    yield {
+      textureSize: [modifiedWidth, textureSize[1], textureSize[2]],
+    };
+  }
+  if (info.blockHeight! > 1 && modifiedHeight !== textureSize[1]) {
+    yield {
+      textureSize: [textureSize[0], modifiedHeight, textureSize[2]],
+    };
+  }
+  if (
+    info.blockWidth! > 1 &&
+    info.blockHeight! > 1 &&
+    modifiedWidth !== textureSize[0] &&
+    modifiedHeight !== textureSize[1]
+  ) {
+    yield {
+      textureSize: [modifiedWidth, modifiedHeight, textureSize[2]],
+    };
+  }
+}
+
+// Test that copying various mip levels works.
+// Covers two special code paths:
+//    the physical size of the subresouce is not equal to the logical size
+//    bufferSize - offset < bytesPerImage * copyExtent.depth and copyExtent needs to be clamped for all BC formats
+g.test('copy_various_mip_levels')
+  .params(
+    params()
+      .combine(kMethodsToTest)
+      .combine([
+        {
+          copySizeInBlocks: { width: 5, height: 4, depth: 1 },
+          originInBlocks: { x: 3, y: 2, z: 0 },
+          texturePhysicalSizeAtMipLevelInBlocks: { width: 8, height: 6, depth: 1 },
```

Comment:
nit: IIUC it looks like this could be a private param (`_texturePhysicalSizeAtMipLevelInBlocks`) since the actual identifying part of the params is `textureSize`